pkg/scheduler/factory.go

```diff
@@ -357,21 +357,21 @@ func MakeDefaultErrorFunc(client clientset.Interface, podLister corelisters.PodL
 		// Check if the Pod exists in informer cache.
 		cachedPod, err := podLister.Pods(pod.Namespace).Get(pod.Name)
 		if err != nil {
-			klog.InfoS("Pod doesn't exist in informer cache", "pod", klog.KObj(pod), "err", err)
-			return
+			return fmt.Errorf("obtaining pod from cache: %w", err)
```

Comment:
Technically that's possible but quite rare. To keep it 100% away from the timing issue, we may have to verify if the pod exists inside nominatedPodMap#AddNominatedPod(), but sounds too much?

Reply:
> Technically that's possible but quite rare. To keep it 100% away from the timing issue, we may have to verify if the pod exists inside nominatedPodMap#AddNominatedPod(), but sounds too much?

A race condition is a race condition, addressing that is never "too much" :) The original issue we are addressing is also "rare", but at scale becomes noticeable. We need to make sure that the pod exists while holding the queue lock.

Reply:
I compared the code with v1.18. In 1.18, we sort of immediately add the nominated node info (in preempt()), but it also has chances to be racing with a PodDelete event. (if the delete event occurs [after L410 and before L421](https://github.com/kubernetes/kubernetes/blob/v1.18.19/pkg/scheduler/scheduler.go#L410-L421)).

While since 1.19, we delayed the action to update the nominated node intentionally (#91535) to save one API update operation, and hence the racing issue is more noticeable.

To absolutely eliminate the racing issue, I revised the solution to introduce a `SafePodNominator` and add a UT. @ahg-g @alculquicondor PTAL.

Reply:
Thanks Wei, sounds good to me, but I agree with Aldo, lets make it straightforward and pass in the lister.